File: api-clients/elixir/lib/otreeba_open_cannabis_api/api/seedCompanies.ex
defmodule OtreebaOpenCannabisAPI.Api.SeedCompanies do
  @moduledoc """
  Documentation for OtreebaOpenCannabisAPI.Api.SeedCompanies.
  """

  use Tesla

  plug Tesla.Middleware.BaseUrl, "https://api.otreeba.com/v1"
  plug Tesla.Middleware.JSON

  @doc """
  Get a list of all current seed companies.

  Returns a paginated list of seed companies.
  """
  def get_seed_companies(page, count, sort) do
    method = [method: :get]
    url = [url: "/seed-companies"]
    query_params = [query: [{:"page", page}, {:"count", count}, {:"sort", sort}]]
    header_params = []
    body_params = []
    form_params = []
    params = query_params ++ header_params ++ body_params ++ form_params
    opts = []
    options = method ++ url ++ params ++ opts

    request(options)
  end

  @doc """
  Find seed company by Open Cannabis Product Code (OCPC).

  Returns a single seed company.
  """
  def get_seed_company_by_ocpc(ocpc) do
    method = [method: :get]
    url = [url: "/seed-companies/#{ocpc}"]
    query_params = []
    header_params = []
    body_params = []
    form_params = []
    params = query_params ++ header_params ++ body_params ++ form_params
    opts = []
    options = method ++ url ++ params ++ opts

    request(options)
  end

  @doc """
  Find strains for a seed company by Open Cannabis Product Code (OCPC).

  Returns a paginated list of strains for a single seed company.
  """
  def get_seed_company_strains_by_ocpc(ocpc, page, count) do
    method = [method: :get]
    url = [url: "/seed-companies/#{ocpc}/strains"]
    query_params = [query: [{:"page", page}, {:"count", count}]]
    header_params = []
    body_params = []
    form_params = []
    params = query_params ++ header_params ++ body_params ++ form_params
    opts = []
    options = method ++ url ++ params ++ opts

    request(options)
  end
end

File: api-clients/clojure/src/otreeba_open_cannabis_api/api/seed_companies.clj
(ns otreeba-open-cannabis-api.api.seed-companies
  (:require [otreeba-open-cannabis-api.core :refer [call-api check-required-params with-collection-format]])
  (:import (java.io File)))

(defn get-seed-companies-with-http-info
  "Get a list of all current seed companies.
  Returns a paginated list of seed companies."
  ([] (get-seed-companies-with-http-info nil))
  ([{:keys [page count sort ]}]
   (call-api "/seed-companies" :get
             {:path-params   {}
              :header-params {}
              :query-params  {"page" page "count" count "sort" sort }
              :form-params   {}
              :content-types ["application/json"]
              :accepts       ["application/json"]
              :auth-names    []})))

(defn get-seed-companies
  "Get a list of all current seed companies.
  Returns a paginated list of seed companies."
  ([] (get-seed-companies nil))
  ([optional-params]
   (:data (get-seed-companies-with-http-info optional-params))))

(defn get-seed-company-by-ocpc-with-http-info
  "Find seed company by Open Cannabis Product Code (OCPC).
  Returns a single seed company."
  [ocpc ]
  (call-api "/seed-companies/{ocpc}" :get
            {:path-params   {"ocpc" ocpc }
             :header-params {}
             :query-params  {}
             :form-params   {}
             :content-types ["application/json"]
             :accepts       ["application/json"]
             :auth-names    []}))

(defn get-seed-company-by-ocpc
  "Find seed company by Open Cannabis Product Code (OCPC).
  Returns a single seed company."
  [ocpc ]
  (:data (get-seed-company-by-ocpc-with-http-info ocpc)))

(defn get-seed-company-strains-by-ocpc-with-http-info
  "Find strains for a seed company by Open Cannabis Product Code (OCPC).
  Returns a paginated list of strains for a single seed company."
  ([ocpc ] (get-seed-company-strains-by-ocpc-with-http-info ocpc nil))
  ([ocpc {:keys [page count ]}]
   (call-api "/seed-companies/{ocpc}/strains" :get
             {:path-params   {"ocpc" ocpc }
              :header-params {}
              :query-params  {"page" page "count" count }
              :form-params   {}
              :content-types ["application/json"]
              :accepts       ["application/json"]
              :auth-names    []})))

(defn get-seed-company-strains-by-ocpc
  "Find strains for a seed company by Open Cannabis Product Code (OCPC).
  Returns a paginated list of strains for a single seed company."
  ([ocpc ] (get-seed-company-strains-by-ocpc ocpc nil))
  ([ocpc optional-params]
   (:data (get-seed-company-strains-by-ocpc-with-http-info ocpc optional-params))))

